File: RGLockbox/RGLockbox.m
/* Copyright (c) 01/27/2016, Ryan Dignard
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 
 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */

#import "RGLockbox.h"
#import <Security/Security.h>
#import <objc/runtime.h>

#if TARGET_OS_IOS || TARGET_OS_TV
    #import <UIKit/UIKit.h>
#endif

#pragma mark - Swizzle
static void rg_swizzle(Class RG_SUFFIX_NULLABLE cls, SEL RG_SUFFIX_NULLABLE original, SEL RG_SUFFIX_NULLABLE replace) {
    IMP replacementImp = method_setImplementation(class_getInstanceMethod(cls, replace),
                                                  class_getMethodImplementation(cls, original));
    // get the replacement IMP
    // we assume swizzle is called on the class with replacement, so we can safety force original onto replacement
    // set the original IMP on the replacement selector
    // try to add the replacement IMP directly to the class on original selector
    // if it succeeds then we're all good (the original before was located on the superclass)
    // if it doesn't then that means an IMP is already there so we have to overwrite it
    method_setImplementation(class_getInstanceMethod(cls, original), replacementImp);
}

#pragma mark - Bundle Identifier
static NSString* _sBundleIdentifier;
static NSString* RG_SUFFIX_NONNULL override_rg_bundle_identifier(void) {
    return _sBundleIdentifier;
}

static NSString* RG_SUFFIX_NONNULL backing_rg_bundle_identifier(void) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sBundleIdentifier = [NSBundle bundleForClass:[RGLockbox self]].infoDictionary[(id)kCFBundleIdentifierKey];
        rg_bundle_identifier = override_rg_bundle_identifier;
    });
    return _sBundleIdentifier;
}
NSString* RG_SUFFIX_NONNULL (* RG_SUFFIX_NONNULL rg_bundle_identifier)(void) = backing_rg_bundle_identifier;

static RGMultiStringKey* RG_SUFFIX_NONNULL rg_multi_key(NSString* RG_SUFFIX_NULLABLE nameSpace,
                                                        NSString* RG_SUFFIX_NULLABLE key,
                                                        NSString* RG_SUFFIX_NULLABLE accountName,
                                                        NSString* RG_SUFFIX_NULLABLE accessGroup) {
    RGMultiStringKey* ret = [RGMultiStringKey new];
    ret.first = nameSpace ? [NSString stringWithFormat:@"%@.%@", nameSpace, key] : key;
    ret.second = accountName;
    ret.third = accessGroup;
    return ret;
}

static NSMutableDictionary* RG_SUFFIX_NONNULL rg_generic_query(RGMultiStringKey* RG_SUFFIX_NULLABLE key) {
    NSMutableDictionary* query = [@{
                                    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
                                    rg_synchronizable_key() : rg_synchronizable_any()
                                    } mutableCopy];
    if (key.first) {
        query[(__bridge id)kSecAttrService] = key.first;
    }
    if (key.second) {
        query[(__bridge id)kSecAttrAccount] = key.second;
    }
    if (key.third) {
        query[rg_accessgroup_key()] = key.third;
    }
    return query;
}

#pragma mark - Global Symbols
OSStatus (* RG_SUFFIX_NONNULL rg_SecItemCopyMatch)(CFDictionaryRef RG_SUFFIX_NONNULL,
                                                   CFTypeRef* RG_SUFFIX_NULLABLE) = &SecItemCopyMatching;
OSStatus (* RG_SUFFIX_NONNULL rg_SecItemAdd)(CFDictionaryRef RG_SUFFIX_NONNULL,
                                             CFTypeRef RG_SUFFIX_NULLABLE * RG_SUFFIX_NULLABLE) = &SecItemAdd;
OSStatus (* RG_SUFFIX_NONNULL rg_SecItemDelete)(CFDictionaryRef RG_SUFFIX_NONNULL) = &SecItemDelete;

NSString* RG_SUFFIX_NONNULL RGApplicationWillResignActive;
NSString* RG_SUFFIX_NONNULL RGApplicationWillBackground;
NSString* RG_SUFFIX_NONNULL RGApplicationWillTerminate;

#pragma mark - RGLockbox Implementation
@implementation RGLockbox

+ (void) load {
#if TARGET_OS_IOS || TARGET_OS_TV
    RGApplicationWillResignActive = UIApplicationWillResignActiveNotification;
    RGApplicationWillBackground = UIApplicationDidEnterBackgroundNotification;
    RGApplicationWillTerminate = UIApplicationWillTerminateNotification;
#elif TARGET_OS_WATCH
    RGApplicationWillResignActive = @"UIApplicationWillResignActiveNotification";
    RGApplicationWillBackground = @"UIApplicationDidEnterBackgroundNotification";
    RGApplicationWillTerminate = @"UIApplicationWillTerminateNotification";
#elif TARGET_OS_MAC
    RGApplicationWillResignActive = @"NSApplicationWillResignActiveNotification";
    RGApplicationWillBackground = @"NSApplicationWillHideNotification";
    RGApplicationWillTerminate = @"NSApplicationWillTerminateNotification";
#else
#warning "Unknown platform target"
    RGApplicationWillResignActive = @"RGApplicationWillResignActive";
    RGApplicationWillBackground = @"RGApplicationWillBackground";
    RGApplicationWillTerminate = @"RGApplicationWillTerminate";
#endif
}

+ (void) initialize {
    [super initialize];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(flushQueue:)
                                                 name:RGApplicationWillResignActive
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(flushQueue:)
                                                 name:RGApplicationWillBackground
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(flushQueue:)
                                                 name:RGApplicationWillTerminate
                                               object:nil];
}

+ (void) flushQueue:(NSNotification*)notification {
    RGLogs(kRGLogSeverityTrace, @"flushQueue: called on %@", notification.name);
    dispatch_barrier_sync(self.keychainQueue, ^{});
}

+ (RGLockbox*) manager {
    static dispatch_once_t onceToken;
    static RGLockbox* _sManager;
    dispatch_once(&onceToken, ^{
        _sManager = [RGLockbox new];
    });
    return _sManager;
}

#pragma mark - keychainQueue
static dispatch_queue_t _sQueue;
+ (RG_PREFIX_NONNULL dispatch_queue_t) keychainQueue {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sQueue = dispatch_queue_create("RGLockbox-Sync", DISPATCH_QUEUE_SERIAL);
        rg_swizzle(objc_getMetaClass("RGLockbox"), @selector(keychainQueue), @selector(override_keychainQueue));
    });
    return _sQueue;
}

+ (RG_PREFIX_NONNULL dispatch_queue_t) override_keychainQueue {
    return _sQueue;
}

#pragma mark - valueCacheLock
static NSLock* _sValueCacheLock;
+ (RG_PREFIX_NONNULL NSLock*) valueCacheLock {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sValueCacheLock = [NSLock new];
        rg_swizzle(objc_getMetaClass("RGLockbox"), @selector(valueCacheLock), @selector(override_valueCacheLock));
    });
    return _sValueCacheLock;
}

+ (RG_PREFIX_NONNULL NSLock*) override_valueCacheLock {
    return _sValueCacheLock;
}

#pragma mark - valueCache
static NSMutableDictionary* _sValueCache;
+ (RG_PREFIX_NONNULL NSMutableDictionary RG_GENERIC(RGMultiStringKey*, id) *) valueCache {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sValueCache = [NSMutableDictionary new];
        rg_swizzle(objc_getMetaClass("RGLockbox"), @selector(valueCache), @selector(override_valueCache));
    });
    return _sValueCache;
}

+ (RG_PREFIX_NONNULL NSMutableDictionary RG_GENERIC(RGMultiStringKey*, id) *) override_valueCache {
    return _sValueCache;
}

#pragma mark - Public Methods
- (RG_PREFIX_NONNULL instancetype) init {
    return [self initWithNamespace:rg_bundle_identifier()
                     accessibility:nil
                       accountName:nil
                       accessGroup:nil
                      synchronized:NO];
}

- (RG_PREFIX_NONNULL instancetype) initWithNamespace:(RG_PREFIX_NULLABLE NSString*)nameSpace
                                       accessibility:(RG_PREFIX_NULLABLE CFStringRef)accessibility
                                         accountName:(RG_PREFIX_NULLABLE NSString*)account {
    return [self initWithNamespace:nameSpace
                     accessibility:accessibility
                       accountName:account
                       accessGroup:nil
                      synchronized:NO];
}

- (RG_PREFIX_NONNULL instancetype) initWithNamespace:(RG_PREFIX_NULLABLE NSString*)nameSpace
                                       accessibility:(RG_PREFIX_NULLABLE CFStringRef)accessibility
                                         accountName:(RG_PREFIX_NULLABLE NSString*)account
                                         accessGroup:(RG_PREFIX_NULLABLE NSString*)accessGroup
                                        synchronized:(BOOL)synchronized {
    self = [super init];
    if (self) {
        CFStringRef nonnullAccessibility = accessibility ?: rg_accessibility_default();
        self->_namespace = [nameSpace copy];
        self->_itemAccessibility = nonnullAccessibility;
        self->_accountName = [account copy];
        self->_accessGroup = [accessGroup copy];
        self->_isSynchronized = synchronized;
    }
    return self;
}

- (RG_PREFIX_NULLABLE id) testCacheForKey:(RG_PREFIX_NONNULL NSString*)key {
    [[[self class] valueCacheLock] lock];
    id value = [[self class] valueCache][rg_multi_key(self.namespace, key, self.accountName, self.accessGroup)];
    [[[self class] valueCacheLock] unlock];
    return value;
}

- (RG_PREFIX_NULLABLE NSData*) dataForKey:(RG_PREFIX_NONNULL NSString*)key {
    RGMultiStringKey* fullKey = rg_multi_key(self.namespace, key, self.accountName, self.accessGroup);
    [[[self class] valueCacheLock] lock];
    id value = [[self class] valueCache][fullKey];
    if (value) {
        [[[self class] valueCacheLock] unlock];
        return [value isKindOfClass:[NSData self]] ? value : nil;
    }
    __block CFTypeRef data = nil;
    __unused __block OSStatus status;
    dispatch_sync([[self class] keychainQueue], ^{
        NSMutableDictionary* query = rg_generic_query(fullKey);
        query[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitOne;
        query[(__bridge id)kSecReturnData] = @YES;
        status = rg_SecItemCopyMatch((__bridge CFDictionaryRef)query, &data);
        RGLogs(kRGLogSeverityTrace, @"SecItemCopyMatching with %@ returned %@", query, @(status));
    });
    NSData* bridgedData = (__bridge_transfer NSData*)data;
    if (status != errSecInteractionNotAllowed) { /* Not allowed means we need to try again so don't cache NSNull */
        [[self class] valueCache][fullKey] = bridgedData ?: [NSNull null];
    } /* NSNull is a placeholder in the cache to say we've tried */
    [[[self class] valueCacheLock] unlock];
    return bridgedData;
}

- (RG_PREFIX_NONNULL NSArray RG_GENERIC(NSString *) *) allItems {
    RGMultiStringKey* fullKey = rg_multi_key(nil, nil, self.accountName, self.accessGroup);
    NSString* nameSpace = self.namespace;
    NSMutableArray RG_GENERIC(NSString *) * output = [NSMutableArray new];
    [[[self class] valueCacheLock] lock];
    __block CFTypeRef items = nil;
    dispatch_sync([[self class] keychainQueue], ^{
        NSMutableDictionary* query = rg_generic_query(fullKey);
        query[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitAll;
        query[(__bridge id)kSecReturnAttributes] = @YES;
        query[(__bridge id)kSecReturnData] = @YES;
        OSStatus status = rg_SecItemCopyMatch((__bridge CFDictionaryRef)query, &items);
        RGLogs(kRGLogSeverityTrace, @"SecItemCopyMatching with %@ returned %@", query, @(status));
        NSAssert(status != errSecInteractionNotAllowed, @"Keychain item unavailable, change itemAccessibility");
    });
    NSArray RG_GENERIC(NSDictionary *) * bridgedArray = (__bridge_transfer NSArray*)items;
    for (NSUInteger i = 0; i < bridgedArray.count; i++) {
        RGMultiStringKey *itemKey = [RGMultiStringKey new];
        id service = bridgedArray[i][(__bridge id)kSecAttrService];
        if (service) {
            NSAssert([service isKindOfClass:[NSString self]], @"Wrong type");
            itemKey.first = service;
        }
        id account = bridgedArray[i][(__bridge id)kSecAttrAccount];
        if (account) {
            NSAssert([account isKindOfClass:[NSString self]], @"Wrong type");
            itemKey.second = account;
        }
        id accessGroup = bridgedArray[i][rg_accessgroup_key()];
        if (accessGroup) {
            NSAssert([accessGroup isKindOfClass:[NSString self]], @"Wrong type");
            itemKey.third = accessGroup;
        }
        id data = bridgedArray[i][(__bridge id)kSecValueData];
        [[self class] valueCache][itemKey] = data ?: [NSNull null];
        if (service && !nameSpace) {
            [output addObject:service];
        } else if ([service hasPrefix:nameSpace]) {
            NSRange range = [service rangeOfString:nameSpace];
            [output addObject:[service substringFromIndex:range.location + range.length + 1]];
        }
    }
    [[[self class] valueCacheLock] unlock];
    return output;
}

- (void) setData:(RG_PREFIX_NULLABLE NSData*)object forKey:(RG_PREFIX_NONNULL NSString*)key {
    RGMultiStringKey* fullKey = rg_multi_key(self.namespace, key, self.accountName, self.accessGroup);
    [[[self class] valueCacheLock] lock];
    [[self class] valueCache][fullKey] = object ?: [NSNull null];
    dispatch_async([[self class] keychainQueue], ^{
        NSMutableDictionary* query = rg_generic_query(fullKey);
        OSStatus status = rg_SecItemDelete((__bridge CFDictionaryRef)query);
        RGLogs(kRGLogSeverityTrace, @"SecItemDelete with %@ returned %@", query, @(status));
        NSAssert(status != errSecInteractionNotAllowed, @"Keychain item unavailable, change itemAccessibility");
        if (object) { /* Add it */
            [query addEntriesFromDictionary:@{
                                              (__bridge id)kSecValueData : object,
                                              rg_accessibility_key() : (__bridge id)self.itemAccessibility,
                                              rg_synchronizable_key() : @(self.isSynchronized)
                                              }];
            status = rg_SecItemAdd((__bridge CFDictionaryRef)query, NULL);
            RGLogs(kRGLogSeverityTrace, @"SecItemAdd with %@ returned %@", query, @(status));
            NSAssert(status != errSecInteractionNotAllowed, @"Keychain item unavailable, change itemAccessibility");
        }
    });
    [[[self class] valueCacheLock] unlock];
}

- (NSString*) description {
    return [NSString stringWithFormat:@"%@\n\tnamespace:%@"
                                      @"\n\taccountName:%@"
                                      @"\n\taccessGroup:%@"
                                      @"\n\titemAccessibility:%@"
                                      @"\n\tisSynchronized:%@",
                                      super.description,
                                      self.namespace,
                                      self.accountName,
                                      self.accessGroup,
                                      self.itemAccessibility,
                                      @(self.isSynchronized)];
}

@end

File: RGLockbox/RGLockbox.swift
/* Copyright (c) 02/21/2016, Ryan Dignard
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */

import Foundation
import Security

/**
 block used to retrieve an item from the keychain.  Defaults to `SecItemCopyMatching`.
 */
public var rg_SecItemCopyMatch = { SecItemCopyMatching($0, $1) }

/**
 block used to add a nonexistent item to the keychain.  Defaults to `SecItemAdd`.
 */
public var rg_SecItemAdd = { SecItemAdd($0, nil) }

/**
 block used to delete an item from the keychain.  Defaults to `SecItemDelete`.
 */
public var rg_SecItemDelete = { SecItemDelete($0) }

#if os(iOS) || os(tvOS)
    import UIKit
    public let RGApplicationWillResignActive = UIApplicationWillResignActiveNotification
    public let RGApplicationWillBackground = UIApplicationDidEnterBackgroundNotification
    public let RGApplicationWillTerminate = UIApplicationWillTerminateNotification
#elseif os(watchOS)
    public let RGApplicationWillResignActive = "UIApplicationWillResignActiveNotification"
    public let RGApplicationWillBackground = "UIApplicationDidEnterBackgroundNotification"
    public let RGApplicationWillTerminate = "UIApplicationWillTerminateNotification"
#elseif os(OSX)
    public let RGApplicationWillResignActive = "NSApplicationWillResignActiveNotification"
    public let RGApplicationWillBackground = "NSApplicationWillHideNotification"
    public let RGApplicationWillTerminate = "NSApplicationWillTerminateNotification"
#else
    public let RGApplicationWillResignActive = "RGApplicationWillResignActive"
    public let RGApplicationWillBackground = "RGApplicationWillBackground"
    public let RGApplicationWillTerminate = "RGApplicationWillTerminate"
#endif

/**
 Instances of RGLockbox manage access to a given keychain service name.  The default service is your app's bundle identifier.  A given manager is threadsafe.
 */
public class RGLockbox {

/**
 Keychain accesses are performed on this queue to keep the cache in sync with the backing store.
 */
    public static let keychainQueue = DispatchQueue(label: "RGLockbox-Sync")
    
/**
 This lock controls access to `valueCache`.
 */
    static let valueCacheLock = NSLock()
    
/**
 Your app's bundle identifier pre-calculated.
 */
    public static var bundleIdentifier:String? = Bundle.main.infoDictionary?[kCFBundleIdentifierKey as String] as? String
    
/**
 `valueCache` stores in memory the values known to all managers.  A key that has been seen before will used the cached value.
 */
    public static var valueCache:[RGMultiKey : Any] = [:]
    
/**
 Determines the service name used by the manager.
 */
    public let namespace:String?
    
/**
 Determines the accessibility assigned by the manager to a given item on add or update.
 */
    public let itemAccessibility:CFString
    
/**
 Qualifies entries by account if provided.
 */
    public let accountName:String?
    
/**
 Qualifies searches and writes to this accessGroup if provided.
 */
    public let accessGroup:String?
    
/**
 Marks items written by this manager to be synchronizable.
 */
    public let isSynchronized:Bool
    
/**
 Creates a new `RGLockbox` instance with default namespace and item accessibility.
 */
    public class func manager() -> RGLockbox {
        return RGLockbox()
    }
    
/**
 Registers for application state changes at the class level.
 */
    private var onceToken = dispatch_once_t()
    
/**
 A new instance of `RGLockbox`.
 - parameter namespace: The service to which the instance is associated.
 - parameter accessibilty: The item accessibility to write to the keychain items.
 - parameter accountName: The manager's associated account if account qualified.
 - parameter accessGroup: The manager's associated accessGroup if restricted.
 - parameter synchronized: Whether this manager's writes will be marked synchronizable.
 - returns: An instance of `RGLockbox` with the provided namespace and accessibility.
 */
    public required init(withNamespace namespace:String? = RGLockbox.bundleIdentifier,
                                       accessibility:CFString = kSecAttrAccessibleAfterFirstUnlock,
                                       accountName:String? = nil,
                                       accessGroup:String? = nil,
                                       synchronized:Bool = false) {
        dispatch_once(&onceToken) {
            NSNotificationCenter.defaultCenter().addObserverForName(RGApplicationWillResignActive,
                                                                    object: nil,
                                                                    queue: nil,
                                                                    usingBlock: {_ in 
                RGLogs(.Trace, "flushQueue on RGApplicationWillResignActive")
                dispatch_barrier_sync(RGLockbox.keychainQueue, {})
            })
            NSNotificationCenter.defaultCenter().addObserverForName(RGApplicationWillBackground,
                                                                    object: nil,
                                                                    queue: nil,
                                                                    usingBlock: {_ in
                RGLogs(.Trace, "flushQueue on RGApplicationWillBackground")
                dispatch_barrier_sync(RGLockbox.keychainQueue, {})
            })
            NSNotificationCenter.defaultCenter().addObserverForName(RGApplicationWillTerminate,
                                                                    object: nil,
                                                                    queue: nil,
                                                                    usingBlock: {_ in
                RGLogs(.Trace, "flushQueue on RGApplicationWillTerminate")
                dispatch_barrier_sync(RGLockbox.keychainQueue, {})
            })
        }
        self.namespace = namespace
        self.itemAccessibility = accessibility
        self.accountName = accountName
        self.accessGroup = accessGroup
        self.isSynchronized = synchronized
    }
    
/**
 Raw read access to the keychain.  Caches reads to `valueCache`.
 - parameter key: The key used to identify the item.
 - returns: `Data` which is `nil` if not found.
 */
    @discardableResult
    public func dataForKey(_ key:String) -> Data? {
        let name = namespace != nil ? "\(namespace!).\(key)" : key
        let fullKey = RGMultiKey(withFirst: name, second: self.accountName, third: self.accessGroup)
        RGLockbox.valueCacheLock.lock()
        let value = RGLockbox.valueCache[fullKey]
        if value != nil {
            RGLockbox.valueCacheLock.unlock()
            RGLogs(.trace, "returning prematurely for key \(key) and value \(value)")
            return value is Data ? (value as! Data) : nil
        }
        var data:AnyObject? = nil
        RGLockbox.keychainQueue.sync(execute: {
            RGLogs(.trace, "hit sync with key \(key)")
            var query:[NSString:Any] = [
                kSecClass : kSecClassGenericPassword,
                kSecAttrService : fullKey.first!,
                kSecMatchLimit : kSecMatchLimitOne,
                kSecReturnData : true,
                kSecAttrSynchronizable : kSecAttrSynchronizableAny
            ]
            query[kSecAttrAccount] = fullKey.second
            query[kSecAttrAccessGroup] = fullKey.third
            let status = rg_SecItemCopyMatch(query as NSDictionary, &data)
            RGLogs(.trace, "SecItemCopyMatching with \(query) returned \(status)")
        })
        let bridgedData = data as! Data?
        RGLockbox.valueCache[fullKey] = bridgedData != nil ? bridgedData : NSNull()
        RGLockbox.valueCacheLock.unlock()
        return bridgedData
    }
    
    public func allItems() -> Array<String> {
        let fullKey = RGMultiKey(second: self.accountName, third: self.accessGroup)
        var data:AnyObject? = nil
        RGLockbox.valueCacheLock.lock()
        RGLockbox.keychainQueue.sync(execute: {
            RGLogs(.trace, "hit sync with fetch all")
            var query:[NSString:Any] = [
                kSecClass : kSecClassGenericPassword,
                kSecMatchLimit : kSecMatchLimitAll,
                kSecReturnAttributes : true,
                kSecReturnData : true,
                kSecAttrSynchronizable : kSecAttrSynchronizableAny
            ]
            query[kSecAttrAccount] = fullKey.second
            query[kSecAttrAccessGroup] = fullKey.third
            let status = rg_SecItemCopyMatch(query as NSDictionary, &data)
            RGLogs(.trace, "SecItemCopyMatching with \(query) returned \(status)")
        })
        let items = data as? Array<Dictionary<String, Any>>
        var output:Array<String> = []
        for item in items ?? [] {
            let itemKey = RGMultiKey()
            let service = item[kSecAttrService as String]
            if let service = service {
                let serviceName = (service as! String)
                itemKey.first = serviceName
                if self.namespace == nil {
                    output.append(serviceName)
                } else if serviceName.hasPrefix("\(self.namespace!).") {
                    let range = serviceName.rangeOfString("\(self.namespace!).")
                    output.append(serviceName.substringFromIndex(range!.endIndex))
                }
            }
            let account = item[kSecAttrAccount as String]
            if let account = account {
                itemKey.second = (account as! String)
            }
            let accessGroup = item[kSecAttrAccessGroup as String]
            if let accessGroup = accessGroup {
                itemKey.third = (accessGroup as! String)
            }
            let data = item[kSecValueData as String] as? NSData
            RGLockbox.valueCache[itemKey] = data != nil ? data : NSNull()
        }
        RGLockbox.valueCacheLock.unlock()
        return output
    }
    
/**
 Raw write access to keychain.  Caches writes to `valueCache`.
 - parameter data: The data to store on the given key.  If `nil` clears the value in the keychain.
 - parameter key: The identifier of the keychain item.
 */
    public func setData(_ data:Data?, forKey key:String) {
        let name = namespace != nil ? "\(namespace!).\(key)" : key
        let fullKey = RGMultiKey(withFirst: name, second: self.accountName, third: self.accessGroup)
        RGLockbox.valueCacheLock.lock()
        RGLockbox.valueCache[fullKey] = ((data != nil) ? data : NSNull())
        RGLockbox.keychainQueue.async(execute: {
            RGLogs(.trace, "key is \(fullKey.first) with data \(data)")
            var query:[NSString:Any] = [
                kSecClass : kSecClassGenericPassword,
                kSecAttrService : fullKey.first!,
                kSecAttrSynchronizable : kSecAttrSynchronizableAny
            ]
            query[kSecAttrAccount] = fullKey.second
            query[kSecAttrAccessGroup] = fullKey.third
            var status = rg_SecItemDelete(query as NSDictionary)
            RGLogs(.trace, "SecItemDelete with \(query) returned \(status)")
            assert(status != errSecInteractionNotAllowed, "Keychain item unavailable, change itemAccessibility")
            if let data = data {
                query[kSecValueData] = data
                query[kSecAttrAccessible] = self.itemAccessibility
                query[kSecAttrSynchronizable] = self.isSynchronized
                status = rg_SecItemAdd(query as NSDictionary)
                RGLogs(.trace, "SecItemAdd with \(query) returned \(status)")
                assert(status != errSecInteractionNotAllowed, "Keychain item unavailable, change itemAccessibility")
            }
        })
        RGLockbox.valueCacheLock.unlock()
    }
}
